File: src/main/java/eladkay/quaritum/common/compat/mt/DiagramMT.java
package eladkay.quaritum.common.compat.mt;

import com.google.common.collect.Lists;
import eladkay.quaritum.api.rituals.IDiagram;
import eladkay.quaritum.api.rituals.PositionedBlock;
import eladkay.quaritum.api.rituals.PositionedBlockChalk;
import eladkay.quaritum.api.rituals.RitualRegistry;
import minetweaker.IUndoableAction;
import minetweaker.MineTweakerAPI;
import minetweaker.api.item.IItemStack;
import minetweaker.api.minecraft.MineTweakerMC;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import stanhebben.zenscript.annotations.ZenClass;
import stanhebben.zenscript.annotations.ZenMethod;

import java.util.List;

@ZenClass("mods.quaritum.Blueprint")
public class DiagramMT {
    @ZenMethod
    public static void addRecipe(String name, IItemStack[] input, IItemStack output, boolean onPlayers, int drain, int rarity, int[][] chalks) {
        MineTweakerAPI.apply(new Add(new DiagramCrafting(name, CraftTweaker.getStacks(input), MineTweakerMC.getItemStack(output), getListOfChalks(chalks), drain, onPlayers, rarity, drain > 0)));
    }
    private static List<PositionedBlock> getListOfChalks(int[][] chalks) {
        List<PositionedBlock> ret = Lists.newArrayList();
        for(int[] chal : chalks)
            ret.add(new PositionedBlockChalk(EnumDyeColor.values()[chal[0]], new BlockPos(chal[1], chal[2], chal[3])));
        return ret;
    }

    private static class Add implements IUndoableAction {
        private final DiagramCrafting recipe;

        public Add(DiagramCrafting recipe) {
            this.recipe = recipe;
        }

        @Override
        public void apply() {
            RitualRegistry.registerDiagram(recipe, recipe.getUnlocalizedName());
        }

        @Override
        public boolean canUndo() {
            return true;
        }

        @Override
        public void undo() {
            for (IDiagram r : RitualRegistry.getDiagramList()) {
                if (r.getUnlocalizedName().equals(recipe.getUnlocalizedName())) {
                    RitualRegistry.getDiagramList().remove(r);
                    break;
                }
            }

        }

        @Override
        public String describe() {
            return "Adding Diagram " + recipe.getUnlocalizedName();
        }

        @Override
        public String describeUndo() {
            return "Removing Diagram " + recipe.getUnlocalizedName();
        }

        @Override
        public Object getOverrideKey() {
            return null;
        }
    }

    @ZenMethod
    public static void removeRecipe(IItemStack output) {
        MineTweakerAPI.apply(new Remove(MineTweakerMC.getItemStack(output)));
    }

    private static class Remove implements IUndoableAction {
        private final ItemStack output;
        private IDiagram recipe;

        public Remove(ItemStack output) {
            this.output = output;
        }

        @Override
        public void apply() {
            for (IDiagram r : RitualRegistry.getDiagramList()) {
                if (r.getUnlocalizedName().equals(recipe.getUnlocalizedName())) {
                    recipe = r;
                    break;
                }
            }

            RitualRegistry.getDiagramList().remove(recipe);
        }

        @Override
        public boolean canUndo() {
            return recipe != null;
        }

        @Override
        public void undo() {
            RitualRegistry.registerDiagram(recipe, recipe.getUnlocalizedName());
        }

        @Override
        public String describe() {
            return "Removing Diagram Recipe for " + output.getDisplayName();
        }

        @Override
        public String describeUndo() {
            return "Restoring Diagram Recipe for " + output.getDisplayName();
        }

        @Override
        public Object getOverrideKey() {
            return null;
        }
    }
}

File: src/main/java/eladkay/quaeritum/common/compat/mt/DiagramMT.kt
package eladkay.quaeritum.common.compat.mt

import com.google.common.collect.Lists
import eladkay.quaeritum.api.rituals.IDiagram
import eladkay.quaeritum.api.rituals.PositionedBlock
import eladkay.quaeritum.api.rituals.PositionedBlockChalk
import eladkay.quaeritum.api.rituals.RitualRegistry
import minetweaker.IUndoableAction
import minetweaker.MineTweakerAPI
import minetweaker.api.item.IItemStack
import minetweaker.api.minecraft.MineTweakerMC
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import stanhebben.zenscript.annotations.ZenClass
import stanhebben.zenscript.annotations.ZenMethod

@ZenClass("mods.quaeritum.Blueprint")
object DiagramMT {
    @ZenMethod
    @JvmStatic
    fun addRecipe(name: String, input: Array<IItemStack>, output: IItemStack, onPlayers: Boolean, drain: Int, rarity: Int, chalks: Array<IntArray>) {
        MineTweakerAPI.apply(Add(DiagramCrafting(name, CraftTweaker.getStacks(input), MineTweakerMC.getItemStack(output), getListOfChalks(chalks), drain, onPlayers, rarity, drain > 0)))
    }

    private fun getListOfChalks(chalks: Array<IntArray>): List<PositionedBlock> {
        val ret = Lists.newArrayList<PositionedBlock>()
        for (chal in chalks)
            ret.add(PositionedBlockChalk(EnumDyeColor.values()[chal[0]], BlockPos(chal[1], chal[2], chal[3])))
        return ret
    }

    @ZenMethod
    fun removeRecipe(output: IItemStack) {
        MineTweakerAPI.apply(Remove(MineTweakerMC.getItemStack(output)))
    }

    private class Add(private val recipe: DiagramCrafting) : IUndoableAction {

        override fun apply() {
            RitualRegistry.registerDiagram(recipe, recipe.unlocalizedName)
        }

        override fun canUndo(): Boolean {
            return true
        }

        override fun undo() {
            for (r in RitualRegistry.getDiagramList()) {
                if (r.unlocalizedName == recipe.unlocalizedName) {
                    RitualRegistry.getDiagramList().remove(r)
                    break
                }
            }

        }

        override fun describe(): String {
            return "Adding Diagram " + recipe.unlocalizedName
        }

        override fun describeUndo(): String {
            return "Removing Diagram " + recipe.unlocalizedName
        }

        override fun getOverrideKey(): Any? {
            return null
        }
    }

    private class Remove(private val output: ItemStack) : IUndoableAction {
        private var recipe: IDiagram? = null

        override fun apply() {
            for (r in RitualRegistry.getDiagramList()) {
                if (r.unlocalizedName == recipe!!.unlocalizedName) {
                    recipe = r
                    break
                }
            }

            RitualRegistry.getDiagramList().remove(recipe)
        }

        override fun canUndo(): Boolean {
            return recipe != null
        }

        override fun undo() {
            RitualRegistry.registerDiagram(recipe, recipe!!.unlocalizedName)
        }

        override fun describe(): String {
            return "Removing Diagram Recipe for " + output.displayName
        }

        override fun describeUndo(): String {
            return "Restoring Diagram Recipe for " + output.displayName
        }

        override fun getOverrideKey(): Any? {
            return null
        }
    }
}
